tests/callbacks/test_device_stats_monitor.py

```diff
@@ -0,0 +1,139 @@
+from tests.helpers import BoringModel
+from tests.helpers.runif import RunIf
+
+
+@RunIf(min_torch="1.8")
+@RunIf(min_gpus=1)
+def test_device_stats_gpu_from_torch(tmpdir):
+    """Test GPU stats are logged using a logger with Pytorch >= 1.8.0."""
+    model = BoringModel()
+    device_stats = DeviceStatsMonitor()
+    logger = CSVLogger(tmpdir)
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=2,
+        limit_train_batches=7,
+        log_every_n_steps=1,
+        gpus=1,
+        callbacks=[device_stats],
+        logger=logger,
+        checkpoint_callback=False,
+        enable_progress_bar=False,
+    )
+
+    # This line is needed to initialize cuda context, so that `torch.cuda.memory_stats` returns something.
+    # See https://github.com/pytorch/pytorch/issues/65793
+    _ = torch.randn(1, device="cuda")
+    with patch.object(CSVLogger, "log_metrics") as mock_log_method:
+        device_stats.on_train_batch_start(trainer, model, None, None, None)
+
+    metrics_dict = mock_log_method.call_args.args[0]
+
+    fields = ["allocated_bytes.all.freed", "inactive_split.all.peak", "reserved_bytes.large_pool.peak"]
+
+    for f in fields:
+        assert any(f in h for h in metrics_dict.keys())
+
+
+@RunIf(max_torch="1.7")
+@RunIf(min_gpus=1)
+def test_device_stats_gpu_from_nvidia(tmpdir):
+    """Test GPU stats are logged using a logger with Pytorch < 1.8.0."""
+    model = BoringModel()
+    device_stats = DeviceStatsMonitor()
+    logger = CSVLogger(tmpdir)
+    log_every_n_steps = 2
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=2,
+        limit_train_batches=7,
+        log_every_n_steps=log_every_n_steps,
+        gpus=1,
+        callbacks=[device_stats],
+        logger=logger,
+        checkpoint_callback=False,
+        enable_progress_bar=False,
+    )
+
+    with patch.object(CSVLogger, "log_metrics") as mock_log_method:
+        device_stats.on_train_batch_start(trainer, model, None, None, None)
+
+    metrics_dict = mock_log_method.call_args.args[0]
+
+    fields = ["utilization.gpu", "memory.used", "memory.free", "utilization.memory"]
+
+    for f in fields:
+        assert any(f in h for h in metrics_dict.keys())
+
+
+@RunIf(tpu=True)
+def test_device_stats_monitor_tpu(tmpdir):
+    """Test TPU stats are logged using a logger."""
+
+    model = BoringModel()
+    device_stats = DeviceStatsMonitor()
+    logger = CSVLogger(tmpdir)
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=2,
+        limit_train_batches=5,
+        tpu_cores=8,
+        callbacks=[device_stats],
+        logger=logger,
+        checkpoint_callback=False,
+        enable_progress_bar=False,
+    )
+
+    with patch.object(CSVLogger, "log_metrics") as mock_log_method:
+        device_stats.on_train_batch_start(trainer, model, None, None, None)
+
+    metrics_dict = mock_log_method.call_args.args[0]
```

Comment:
the test is failing with
```
>       metrics_dict = mock_log_method.call_args.args[0]
E       AttributeError: 'NoneType' object has no attribute 'args
```

Reply:
thanks, pushed a fix. at least I hope it's a fix, I have no way of checking by testing locally - is the best way just to push to CI and check there?

Reply:
>  is the best way just to push to CI and check there?

no, because the CI takes 40 mins to finish completely and so you will wait a while.

You probably want to bring the instantiation of the logger under the with block? Is that the problem maybe?
If all does not work, maybe just try this simple way:

```python
logger = CSVLogger()
logger.some_method = Mock()

...

assert logger.some_method....
```

Reply:
the reason I cant test locally is because it is a TPU test, so I have to test with CI right? 

I dont think thats the problem, because the other GPU tests are working just fine and they have a similar structure :/

Reply:
ah ok sorry. 

the error happens here. 
```bash

        with patch.object(CSVLogger, "log_metrics") as mock_log_method:
>           device_stats.on_train_batch_start(trainer, model, None, None, None)

tests/callbacks/test_device_stats_monitor.py:113: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pytorch_lightning/callbacks/device_stats_monitor.py:56: in on_train_batch_start
    device_stats = trainer.accelerator.get_device_stats(device)
pytorch_lightning/accelerators/tpu.py:74: in get_device_stats
    memory_info = xm.get_memory_info(device)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

device = device(type='xla', index=1)

    def get_memory_info(device):
      """Retrieves the device memory information.
    
      Args:
        device (string): The device whose memory information are requested.
    
      Returns:
        A dictionary with `kb_free` (free memory in KB) and `kb_total` (total
        memory in KB) keys.
      """
>     return torch_xla._XLAC._xla_memory_info(str(device))
E     RuntimeError: tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:1604 : Check failed: session->session()->Run({cached_node.outputs[0]}, &outputs) == ::tensorflow::Status::OK() (Aborted: The same RunStep (Master) request was received twice. session_handle: "96af2b0710240929" fetch: "XRTMemoryInfo:0" options { } store_errors_in_response_body: true request_id: -2872753854656592607 vs. OK)
E     *** Begin stack trace ***
E     	tensorflow::CurrentStackTrace()
E     	xla::XrtComputationClient::GetMemoryInfo(std::string const&)
E     	
E 
```

the tpu accelerator probably needs to initialize first before you can access the device. 
you could try to mock the xm module by adding.

`@mock.patch("pytorch_lightning.plugins.training_type.tpu_spawn.xm")`

Reply:
ok will try that, thanks!!

Reply:
no luck :/ It seems we can't mock the xm module. It is super frustrating to debug like this, having to wait for the CI

Reply:
so xla model can't operate on the mocked device. which would mean mocking trainer.accelerator.get_device_stats, which defeats the point of the assertion later on.

if this is meant to be an integration test:
- we could mock trainer.accelerator.get_device_stats to return some dummy data
- and mock trainer.logger.log_metrics to confirm the output of one is the input to the other

but beyond that i'm not sure how valuable this test is

Reply:
this is meant to be an integration test, and I do see value in it to test all the parts working together, like [this line](https://github.com/PyTorchLightning/pytorch-lightning/pull/9712/files#diff-24932847c4a40632b78908036e3f9a6cca82e6dd717d2ec50d6168ab45cf49f1R55) for example

> 'm not sure how valuable this test is since we're asserting on the implementation of accelerator.get_device_stats (which is already separately unit tested)

we are also effectively assert that mock_log_method was called, and has call_args, although I could make that more explicit

Reply:
The Error is the Replication Issue, I have added to the TPU Troubleshooting [docs](https://pytorch-lightning.readthedocs.io/en/latest/advanced/tpu.html#troubleshooting)
```
This error is raised when the XLA device is called outside the spawn process. Internally in TPUSpawn Plugin for training on multiple tpu cores, we use XLA's xmp.spawn. Don't use xm.xla_device() while working on Lightning + TPUs!
```

Hence, we can’t call `get_device_stats` method directly